### ISSUE TITLE
Help with Level 7 Logic

### ISSUE THREAD

**Issue by pwintthiri** [nus-cs2030/2122-s2#462]
# Summary
Hi, I'm quite confused about the logic of taking a rest for Servers here. Does each Server have a rest time upon creation? I put restTimes into the Server's constructor when creating new Servers in Simulate7 but I'm not sure if that's the correct idea.

In my Done event, I had an 
if the server's rest time is not 0, then I changed the server's status to busy and changed the next available time of the server to be `server.endTime() + server.getRestTime().get()`. Then I returned a Break event and replaced the server in the shop.

I also tried printing out the server's rest time through `server.getRestTime().get()` and the results are very confusing. The results are after "REST" and I don't know how to interpret this.

Thank you!

## Screenshots (if any): 

![Screenshot 2022-04-21 at 8 55 24 PM](https://user-images.githubusercontent.com/95198407/164462575-a6eed23f-d8bd-4771-b2bc-4469e6bf45cc.png)

**Comment by lyzhou717** [comment]
Hi i think your implementation is correct! I also assigned resttime to a variable and called it in my Done event

if you're a bit confused you're assigning a supplier (like a lambda) that gets a rest time for your server. So if you look in Main7 u can sort of see how the supplier actually gets/calculates the server resttime. So in this case, the server doesn't have an actual resttime on construction, but only when u call the .get() which prompts the supplier to use the method defined in Main7 to give u a restTime.

**Comment by pwintthiri** [comment]
> Hi i think your implementation is correct! I also assigned resttime to a variable and called it in my Done event
> 
> if you're a bit confused you're assigning a supplier (like a lambda) that gets a rest time for your server. So if you look in Main7 u can sort of see how the supplier actually gets/calculates the server resttime. So in this case, the server doesn't have an actual resttime on construction, but only when u call the .get() which prompts the supplier to use the method defined in Main7 to give u a restTime. (so if u call ur .get() at the wrong timing/place, you might end up w wrong rest times!)

Thanks for your reply! Do you happen to know what the rest values mean? Is it the amount of time the Server will rest?

**Comment by Keefetgy** [comment]
@pwintthiri the rest values are positive numbers. When rest value is >0, that indicates it is time for the server to rest and how long the server should be resting for. When it is 0, the server should not be resting.

**Comment by raynang** [comment]
> Do you happen to know what the rest values mean? Is it the amount of time the Server will rest?

Hello! As mentioned above, the rest values are >= 0 and when it is more than 0,  it means that the server will rest for that amount of time. Similar to the service times from level 5, you only call .get() to retrieve the rest time value when it is needed. If not, you could end up calling it too many times. I used Lazy for level 5 but just used Supplier for level 7.

Not sure how you want to implement it, but in my implementation, I checked for the rest time in my DONE event and updated the next available time for servers that needed to rest.

**Comment by Alexius0123** [comment]
Hello! Actually for my implementation, regardless of whether rest time was 0 or not, I made my Done class return a Break class. The difference would be the event time of Break class and the servers assigned. 

If rest time was 0, I would return a Break class of the same event time as the Done class, as well as an available Server.

On the other hand, if rest time was more than zero, I would return a Break class of event time getAvailableTime of the server + rest time. I would also return a busy Server that would only be available in the Break class.

Hope this helps!

**Comment by evelyynnz** [comment]
Hello! I think my logic is similar to urs, but for me I only called restTime.get() if the event was a Rest event. Thus, in my simulate 7, I  created an if condition whereby if the event is a Rest event, i would replace the event with a new Rest event where the supplier.get() is called. For the rest of my simulate classes, rest event would just have a 0 rest time. (I created 2 constructors for this). Hope this helps!
